### PR TITLE
Filter methods

### PR DESCRIPTION
**Pull request by gero3**
This PR is making it possible to filter on methods and properties with the filter functionality in the documentation.

[Try here](https://rawgit.com/gero3/three.js/filterMethods/docs/index.html)

image

 This is a reponse to #13025.

### CODE REVIEW

**Comment by looeee** [comment]
Wow, that was fast! 😅 

Looks like it mainly works really well too. 
One thing though, when the main class is matched, it expands all the sub-links too:

download

**Comment by gero3** [comment]
The sublinks don't show anymore when the page matches.

**Review by haroldiedema — APPROVED** [review]
Very nice addition :smile:

**Comment by donmccurdy** [comment]
This is awesome and I don't want to vote against it based on maintenance convenience — but here, or maybe in a later PR, could we:

a. generate the search index with a build script that scrapes the HTML docs pages, e.g. looking for `<h2 data-type="method">.load()</h2>` or something
b. generate search index and the docs themselves using JSDoc and [documentation.js](https://github.com/documentationjs/documentation) or similar. The JSDoc could be in separate files from the source code if that's preferable.
c. other ideas?

I'd be OK with merging this and dealing with making updates more convenient later.

**Comment by haroldiedema** [comment]
I know we've discussed about this before, but I still vote for option **b**.

It's widely used in the industry and also keeps your documentation up-to-date, since docs must be updated in PR's since they become part of the source.

**Comment by mrdoob** [comment]
We may want to look into [tern-three.js](https://github.com/tylerwbell/tern-threejs).
We already have a file that we currently maintain by hand here:
https://github.com/mrdoob/three.js/blob/dev/editor/js/libs/tern-threejs/threejs.js

**Comment by gero3** [comment]
This is the build script I used: [gist](https://gist.github.com/gero3/e76d8925a72a848c51a133ba0879e100)

It reads the urls from the list.js file and then searches the methods and properties inside the html file.

**Comment by haroldiedema** [comment]
Looks similar to what [tern-threejs](https://github.com/tylerwbell/tern-threejs) does (e.g. https://github.com/tylerwbell/tern-threejs/blob/master/generate.js#L115)

**Comment by mrdoob** [comment]
@haroldiedema Interesting! I thought tern-threejs was getting the data directly from the source somehow.

**Comment by gero3** [comment]
A (cleaned up) generate script has been added to update the list.js file. It extracts all data for examples and api and saves it (linted) in the list.js file.

Does anything need to be fixed further, message away.

File: utils/docs/build.js [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+
+			pages[ page ] = JSONsort( filterItems );
+
+			allPages[ page ] = filterItems;
+
+		}
+
+	}
+	/*
+    for (var pageToCheck in allPages) {
```
Comment (haroldiedema) [inline utils/docs/build.js]: Does this commented code need to stay?

**Review by haroldiedema — APPROVED** [review]
It feels really 'iffy' that we have to "parse" the documentation files, but since we don't have a better alternative, it looks fine to me.

File: utils/docs/build.js (hunk 1 above)
Comment (gero3) [reply utils/docs/build.js]: This code actually adds all methods from basetype to the list.js file too.

Comment (haroldiedema) [reply utils/docs/build.js]: Well, it currently does nothing. :smile: 

If this really is a needed feature, how about adding a command line switch to the script?
For example: `$ node utils/docs/build.js --include-base-type-methods`

**Review by haroldiedema — APPROVED** [review]

File: utils/docs/build.js (hunk 1 above)
Comment (gero3) [reply utils/docs/build.js]: I've added the command line switch.

**Comment by looeee** [comment]
Is this still the correct testing link? https://rawgit.com/gero3/three.js/filterMethods/docs/index.html

Testing with random letter combos, I seem to be getting quite a few false positives. 

untitled-1

untitled-2

**Comment by gero3** [comment]
@looeee, Sorry a refactor went wrong.

It didn't show the filtered members anymore due to [inversion](https://github.com/mrdoob/three.js/pull/13027/commits/4b50bc3edf158d3a395c9fd43d1ef0619e20f3cd).

**Review by looeee — APPROVED** [review]

File: docs/index.html [hunk 2]
```diff
@@ -1,371 +1,39 @@
...
+<head>
+	<meta charset="utf-8">
+	<title>three.js / documentation</title>
+	<meta name="viewport" content="width=device-width, user-scalable=no, minimum-scale=1.0, maximum-scale=1.0">
+	<link type="text/css" rel="stylesheet" href="index.css">
+</head>
+<body>
 
-		<div id="panel" class="collapsed">
+    <div id="panel" class="collapsed">
```
Comment (mrdoob) [inline docs/index.html]: Soft tabs 😇

**Comment by gero3** [comment]
Soft tabs are now removed in this PR

**Comment by gero3** [comment]
Anything holding this PR back??

**Comment by looeee** [comment]
@mrdoob any change of getting this merged? It's a nice addition to the docs.

**Comment by zsavicreative** [comment]
what happened with this?? those of you who are familiar with threejs know what to search to get the info you're after. Some of us who are learning the ropes don't know what keywords will produce the thing we're after. For example searching lerp gives me nothing. 

image


I get a better result using google operator ```lerp site:threejs.org```


image

**Pull request closed without merging**